### ISSUE TITLE
Inconsistent trade entries with custom SL and TP settings

### ISSUE THREAD

**Issue by mllamazares** [freqtrade/freqtrade#9582]
* Operating system: 6.6.7-4-MANJARO
  * Docker image: freqtradeorg/freqtrade:stable_plot
  * Freqtrade Version: freqtrade 2023.11
  
---

## What I Want to Do
I aim to set a SL and TP based on the features of the candle where the trade opens. That is, once the trade is open, the TP or SL won't need updating (no _trailing_). I assume this is a fairly common scenario, so I must have overlooked something.

## What I've Tried
To tackle this task, I have:
1. Created a dictionary called `custom_info` (I'm aware this won't persist [as mentioned in the docs](https://www.freqtrade.io/en/latest/strategy-advanced/#storing-information)).
2. Stored the TP and SL values in the `custom_info` dictionary before the strategy enters the market in `confirm_trade_entry()`.
3. Set up `custom_exit()` and `custom_stoploss()` functions.

```python
    # Stoploss:
    stoploss = -9.9 

    # Trailing stop:
    trailing_stop = False
    # trailing_only_offset_is_reached = False
    # trailing_stop_positive = 0.01
    # trailing_stop_positive_offset = 0.0  # Disabled / not configured

    timeframe = '30m'

    process_only_new_candles = True

    # These values can be overridden in the config.
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False
    use_custom_stoploss = True

    custom_info = {}

    ...

    def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: float,
                            time_in_force: str, current_time: datetime, entry_tag: Optional[str],
                            side: str, **kwargs) -> bool:
        
        dataframe, _ = self.dp.get_analyzed_dataframe(pair=pair,
                                                     timeframe=self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()

        tp = last_candle['bb_upperband']
        sl = last_candle['low'] 

        sl_pct = (last_candle['close'] - sl) / last_candle['close'] 

        self.custom_info[pair] = {
            'tp': tp,
            'sl_pct': sl_pct
        }

        return True

    def custom_exit(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs):

        if current_rate >= self.custom_info[pair]['tp']:
            return 'tp_reached'

    def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime,
                        current_rate: float, current_profit: float, **kwargs) -> float:

        return -self.custom_info[pair]['sl_pct'] 
```

<details>
<summary>config.js file</summary>

```json
{
    "max_open_trades": 3,
    "stake_currency": "USDT",
    "stake_amount": "unlimited",
    "tradable_balance_ratio": 0.99,
    "fiat_display_currency": "EUR",
    "dry_run": true,
    "dry_run_wallet": 1000,
    "cancel_open_orders_on_exit": false,
    "trading_mode": "futures",
    "margin_mode": "isolated",
    "unfilledtimeout": {
        "entry": 10,
        "exit": 10,
        "exit_timeout_count": 0,
        "unit": "minutes"
    },
    "entry_pricing": {
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1,
        "price_last_balance": 0.0,
        "check_depth_of_market": {
            "enabled": false,
            "bids_to_ask_delta": 1
        }
    },
    "exit_pricing":{
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1
    },
    "exchange": {
        "name": "binance",
        "key": "",
        "secret": "",
        "ccxt_config": {},
        "ccxt_async_config": {},
        "pair_whitelist": [
            "BTC/USDT:USDT",
            "ETH/USDT:USDT",
            "ALGO/USDT:USDT",
            "ATOM/USDT:USDT",
            "LTC/USDT:USDT",
            "XRP/USDT:USDT",
            "EOS/USDT:USDT",
            "LINK/USDT:USDT",
            "AVAX/USDT:USDT",
            "XLM/USDT:USDT",
            "AAVE/USDT:USDT",
            "ZEC/USDT:USDT",
            "INJ/USDT:USDT",
            "SOL/USDT:USDT",
            "UNI/USDT:USDT",
            "ETC/USDT:USDT",
            "DOT/USDT:USDT",
            "DASH/USDT:USDT",
            "BNB/USDT:USDT",
            "BCH/USDT:USDT",
            "FTM/USDT:USDT",
            "DOGE/USDT:USDT",
            "DYDX/USDT:USDT",
            "ARB/USDT:USDT",
            "MANA/USDT:USDT",
            "NEAR/USDT:USDT",
            "NEO/USDT:USDT",
            "OP/USDT:USDT",
            "RUNE/USDT:USDT",
            "ROSE/USDT:USDT",
            "ADA/USDT:USDT",
            "SUSHI/USDT:USDT",
            "YFI/USDT:USDT",
            "ZIL/USDT:USDT",
            "XMR/USDT:USDT",
            "ANT/USDT:USDT",
            "1INCH/USDT"
        ],
        "pair_blacklist": [
            "BNB/.*"
        ]
    },
    "pairlists": [
        {
            "method": "StaticPairList",
            "number_assets": 20,
            "sort_key": "quoteVolume",
            "min_value": 0,
            "refresh_period": 1800
        }
    ],
    "telegram": {
        "enabled": false,
        "token": "",
        "chat_id": ""
    },
    "api_server": {
        "enabled": true,
        "listen_ip_address": "0.0.0.0",
        "listen_port": 8080,
        "verbosity": "error",
        "enable_openapi": false,
        "jwt_secret_key": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
        "ws_token": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
        "CORS_origins": [],
        "username": "XXXXXXXXXXXXXXXXX",
        "password": "XXXXXXXXXXXXXXXX"
    },
    "bot_name": "freqtrade",
    "initial_state": "running",
    "force_entry_enable": false,
    "internals": {
        "process_throttle_secs": 5
    }
}
```
</details>

Finally, I ran the following command to generate a plot for further analysis:
```bash
docker-compose run --rm freqtrade plot-dataframe --timeframe 30m --strategy MyCoolStrategy --timerange 20230901-20231010 -p BTC/USDT:USDT
```

## What Does Not Work

The plot shows many `enter_long` icons, but there were no associated trade entries. Here are some examples highlighted in pink:

![image](https://github.com/freqtrade/freqtrade/assets/5461023/5ae60fb1-74fc-40ac-b7b7-cd7d2d4727f8)

I thought this might be because a trade was already open, but, as you can see, it happens from the beginning of the chart.

Finally, there are some trades made at the end of the chart that seem to follow the defined logic (highlighted in green). However, there are other `enter_long` instances (highlighted in pink) that do not have corresponding trade entries.

![image](https://github.com/freqtrade/freqtrade/assets/5461023/6b0ef20c-0aaa-442f-a050-abb21e5863f0)

My `confirm_trade_entry()` function *always* returns True, so I'm puzzled as to why those entries were not triggered.

**Comment by xmatthias** [comment]
be careful with plot-dataframe, it'll plot whatever you give it - and will **not** do a backtest on it's own (that's where you're most likely having the problem with, at least based on the description).

As such, i assume that the trades (or lack thereof) are because there's never been a backtest - so plotting is simply showing what the strategy generates on it's own - using the latest backtest results. 
if that backtest-result was actually generated from THAT particular chart is irrelevant to the plotting module - so if the strategy was modified - it may show old / outdated trades.

---

From a process perspective, i'd recommend to start freqtrade in webserver mode (`freqtrade webserver --config ...`) - then go to the UI, login (if you haven't done so) - start a backtest from there, and directly analyze the chart Interactively.

You can obviously do this in the cli as well by first running backtest, followed by the plot-dataframe command ... but the freqUI chart is usually nicer to look at, and WAY more performant once you move into "longer range" datasets with tons of candles (on top of allowing you to interactively configure the charts / lines you'd like to see).

**Comment by mllamazares** [comment]
Thanks, @xmatthias. You're right - I forgot to rerun the backtesting *before* plotting the dataframe; now, I can see more trades open:

```bash
docker-compose run --rm freqtrade backtesting --strategy MyCoolStrategy --timeframe 30m --timerange 20230101- -p BTC/USDT:USDT --max-open-trades 1
docker-compose run --rm freqtrade plot-dataframe --strategy MyCoolStrategy --timeframe 30m --timerange 20230101- -p BTC/USDT:USDT
```

However, some trades do not appear to follow the defined (simple) logic. As you can see, the SL is not triggered at the low of the entry candle marked with the `enter_long` tag (I've highlighted the line in green). It gets closed above that SL with the tag `trailing_stop_loss`:

![image](https://github.com/freqtrade/freqtrade/assets/5461023/df108ea3-965f-4f8d-a4d2-2dffcbae6b45)

As you can see in my first post, I have set a custom stoploss, and the trailing stop loss is disabled. Any idea of what's going on?

**Comment by xmatthias** [comment]
not sure what you're looking at - but what i see is
* the trade enters after the low
* the trade exits based on trailing stoploss logic (which could also be custom stoploss) - which was almost certainly updated at the candle it updates.

Optically - the distance (high to stop) is roughly the distance of "close to low" - and as you're trailing that stop - the logic seems correct.


keep in mind, the value you're returning in the custom_stoploss callback is the _distance from the current value_.
If you do the math, it'll quite surely be "high - your desired stoploss distance" - as that was update at the candle the stop triggered.

You should instead have your logic be inspired ( _inspired_ - not copied! ) from the [stoploss using indicator](https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-stoploss-using-an-indicator-from-dataframe-example) example ... which stores the absolute value, and then calculates it via `stoploss_from_absolute` helper function at each step - and will hence avoid the trailing scenario you find yourself in.

**Comment by mllamazares** [comment]
Thanks, @xmatthias! I figured out how to address the SL calculation. 🥳 

However, I fail to see how does the TP triggering works:

I set the TP in the `confirm_trade_entry()` function (see my first post), and unlike the SL, it is be a fixed level rather than relative to the current candle. 

In the plot below, the TP is represented by the orange line. When entering the trade, it should take the value of the TP in the opening candle (which I've extended and highlighted in blue). However, instead of closing the trade at the point where this level is hit (see the blue arrow), it actually closes a candle later (see the pink arrow).

![image](https://github.com/freqtrade/freqtrade/assets/5461023/1fe001e8-d2a9-4aa1-b883-85355c836a04)

The [_Custom exit signal_ documentation](https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-exit-signal) states:
> Called for each open trade every throttling iteration (approximately every 5 seconds) until the trade is closed.

Could this be because freqtrade manages the take profit order internally rather than through limit orders on the exchange?

However, I have set both the `entry` and `exit` order types to `limit`:

```python
    order_types = {
        'entry': 'limit',
        'exit': 'limit',
        'stoploss': 'limit',
        'stoploss_on_exchange': False
    }
```

Any thoughts on what I might be missing, or any tips on how to achieve the desired behavior? Thanks again!

**Comment by xmatthias** [comment]
It's not a real take profit (roi, in freqtrade terminology) - but a simple dynamic sell signal.
As such, it'll not trigger at "price X" - but at the price defined by the regular pricing logic (though price is not your problem).


especially during backtesting however - custom_exit is called once per (detail) candle.
it's "current_rate" will be the candle open price - which wasn't above the imaginary blue line.

You can to some degree work around this by using `--timeframe-detail` - but backtesting will remain a simulation with limited detail when it comes to intra-candle movements.

**Comment by mllamazares** [comment]
OK, brilliant! Then, considering the strategy timeframe is 30m, setting the `--timeframe-detail 1m` should do the trick. 

I assume this would not be a issue running it in live mode since it would check every 5 secs. However, I need to take into account that *all* my orders will be market, not limit, right? I assume the `order_types` object justs affects roi and fixed stoploss, not dynamic stop and exit calculations.

**Comment by xmatthias** [comment]
order_types has nothing to do with the trigger.
if it's an entry order, it'll follow the "entry" setting
if it's an exit order, it'll follow the exit setting. if you set both to market, then yes, it'll be market orders.

if you set it to limit, it'll be a limit order - where you MAY control the price via the custom pricing callbacks.

**Comment by mllamazares** [comment]
Okay, what I meant is that instead of using `populate_exit_trend()`, I use `custom_stoploss()` and `custom_exit()`. Thus, there is no way to set the SL or TP levels (as OCO) in the exchange via limit orders, as they will be managed by Freqtrade itself. Correct me if I'm wrong.

**Comment by xmatthias** [comment]
You can for sure use stoploss on exchange as long as you're using custom_stoploss.

OCO / TP orders on exchange are currently not supported.
